cleverhans/experimental/certification/utils.py

```diff
@@ -244,3 +246,82 @@ def minimum_eigen_vector(x, num_steps, learning_rate, vector_prod_fn):
   for _ in range(num_steps):
     x = eig_one_step(x, learning_rate, vector_prod_fn)
   return x
+
+
+def tf_lanczos_smallest_eigval(vector_prod_fn,
+                               matrix_dim,
+                               max_iter=1000,
+                               collapse_tol=1e-12,
+                               dtype=tf.float32):
+  """Computes smallest eigenvector and eigenvalue using Lanczos in pure TF.
+
+  Args:
+    vector_prod_fn: function which takes a vector as an input and returns
+      matrix vector product.
+    matrix_dim: dimentionality of the matrix.
+    max_iter: maximum number of iterations.
+    collapse_tol: tolerance to determine collapse of the Krylov subspace
+    dtype: type of data
+
+  Returns:
+    tuple of (eigenvalue, eigenvector) of smallest eigenvalue and corresponding
+    eigenvector.
+  """
+
+  # alpha will store diagonal elements
+  alpha = tf.TensorArray(dtype, size=1, dynamic_size=True, element_shape=())
+  # beta will store off diagonal elements
+  beta = tf.TensorArray(dtype, size=0, dynamic_size=True, element_shape=())
+  # q will store Krylov space basis
+  q_vectors = tf.TensorArray(
+      dtype, size=1, dynamic_size=True, element_shape=(matrix_dim, 1))
+
+  b = tf.random_normal(shape=(matrix_dim, 1), dtype=dtype)
+  w = b / tf.norm(b)
+
+  # Iteration 0 of Lanczos
+  q_vectors = q_vectors.write(0, w)
+  w_ = vector_prod_fn(w)
+  cur_alpha = tf.reduce_sum(w_ * w)
+  alpha = alpha.write(0, cur_alpha)
+  w_ = w_ - tf.scalar_mul(cur_alpha, w)
+  w_prev = w
+  w = w_
+
+  # Subsequent iterations of Lanczos
+  for i in tf.range(1, max_iter):
+    cur_beta = tf.norm(w)
+    if cur_beta < collapse_tol:
+      # return early if Krylov subspace collapsed
+      break
+
+    w = w / cur_beta
```

Comment:
is it possible for `cur_beta` to be 0?

Reply:
cur_beta > collapse_tol at this point. Added comment explaining this